### PR TITLE
Expose Spin Application Directory to WASI FS

### PR DESCRIPTION
**Pull request by ecumene**
Ref: https://github.com/fermyon/spin/pull/542
Ref: https://github.com/fermyon/spin/issues/528

 - [ ] Handle `log-dir` case

### CODE REVIEW

File: crates/engine/src/lib.rs [hunk 1]
```diff
@@ -330,6 +336,18 @@ impl<T: Default> ExecutionContext<T> {
...
 
+        if component.core.wasm.enable_monitoring {
+            let spin_dir = match dirs::home_dir() {
+                Some(h) => h.join(SPIN_HOME),
+                None => env::current_dir()?,
+            };
+
+            wasi_ctx = wasi_ctx.preopened_dir(
+                Dir::open_ambient_dir(&spin_dir, ambient_authority())?,
+                Path::new("/.spin/"),
```
Comment (ecumene) [inline crates/engine/src/lib.rs:347]: How should this be configured?

File: crates/engine/src/lib.rs [hunk 2]
```diff
@@ -330,6 +336,18 @@ impl<T: Default> ExecutionContext<T> {
...
         }
 
+        if component.core.wasm.enable_monitoring {
+            let spin_dir = match dirs::home_dir() {
+                Some(h) => h.join(SPIN_HOME),
+                None => env::current_dir()?,
+            };
+
+            wasi_ctx = wasi_ctx.preopened_dir(
+                Dir::open_ambient_dir(&spin_dir, ambient_authority())?,
```
Comment (lann) [inline crates/engine/src/lib.rs:346]: This should only mount the current application within ~/.spin, right?

File: crates/engine/src/lib.rs (hunk 1 above)
Comment (lann) [reply crates/engine/src/lib.rs:347]: I think initially it doesn't need to be.

File: crates/engine/src/lib.rs (hunk 2 above)
Comment (lann) [reply crates/engine/src/lib.rs:346]: ...hmmm now that I think about it, maybe I had a different assumption than you here :)

File: crates/engine/src/lib.rs [hunk 3]
```diff
@@ -330,6 +335,21 @@ impl<T: Default> ExecutionContext<T> {
...
         }
 
+        if component.core.wasm.enable_monitoring {
+            let sanitized_label = sanitize(&self.config.label);
+            let app_dir = match dirs::home_dir() {
+                Some(h) => h.join(SPIN_HOME).join(&sanitized_label),
+                None => PathBuf::from(&sanitized_label),
+            };
+
+            std::fs::create_dir_all(&app_dir)?;
```
Comment (lann) [inline crates/engine/src/lib.rs:345]: It would be nice to dedupe this w/ `save_output_to_logs` into an e.g. `fn ensure_app_dir(label: &str) -> Result<PathBuf>`.

I'd be fine doing that in a separate PR though.

**Review by lann — COMMENTED** [review]
Implementation looks good! I'm not sure how I feel about adding another example project; we have so many as it is...

File: crates/engine/src/lib.rs (hunk 3 above)
Comment (ecumene) [reply crates/engine/src/lib.rs:345]: I can do that here, certainly. :)

File: crates/manifest/src/lib.rs [hunk 4]
```diff
@@ -158,6 +158,8 @@ pub struct WasmConfig {
     pub mounts: Vec<DirectoryMount>,
     /// Optional list of HTTP hosts the component is allowed to connect.
     pub allowed_http_hosts: Vec<String>,
+    /// Flag to enable access to the Spin home directory
+    pub enable_monitoring: bool,
```
Comment (itowlson) [inline crates/manifest/src/lib.rs:162]: The name seems out of sync with the documentation...?  (I don't fully understand the implementation though so I'm probably missing something!)

Comment (ecumene) [reply crates/manifest/src/lib.rs:162]: No you're not missing anything, I've been too lazy and haven't updated the SIP

Comment (ecumene) [reply crates/manifest/src/lib.rs:162]: All updated now :) It should be nice and clear.

File: crates/engine/src/lib.rs [hunk 5]
```diff
@@ -254,15 +271,16 @@ impl<T: Default> ExecutionContext<T> {
...
         let log_dir = match &self.config.log_dir {
             Some(l) => l.clone(),
-            None => match dirs::home_dir() {
-                Some(h) => h.join(SPIN_HOME).join(&sanitized_label).join("logs"),
-                None => PathBuf::from(&sanitized_label).join("logs"),
-            },
+            None => {
+                let app_dir = self.ensure_app_dir()?;
+                let log_dir = app_dir.join("log");
+                std::fs::create_dir_all(&log_dir)?;
```
[the remark this answers is not in the document]
Comment (ecumene) [reply crates/engine/src/lib.rs:281]: Good catch. I think I moved that like around a couple of times then put it back in the wrong place.

**Comment by lann** [comment]
LGTM; since we didn't get much discussion on the SIP I'd like others to review.

**Review by dicej — APPROVED** [review]
LGTM.  I nitpicked passing an explicit reference to the `trace!` macro, but it's certainly not a blocker.

File: crates/engine/src/lib.rs [hunk 6]
```diff
@@ -330,6 +342,16 @@ impl<T: Default> ExecutionContext<T> {
                 wasi_ctx.preopened_dir(Dir::open_ambient_dir(host, ambient_authority())?, guest)?;
         }
 
+        if component.core.wasm.enable_monitoring {
+            let app_dir = self.app_dir();
+            std::fs::create_dir_all(&app_dir)?;
+            log::trace!("Pre-opening app dir {:?}", &app_dir);
```
[the remark this answers is not in the document]
Comment (ecumene) [reply crates/engine/src/lib.rs]: Good point!

File: crates/manifest/src/lib.rs (hunk 4 above)
Comment (itowlson) [reply crates/manifest/src/lib.rs:162]: @ecumene just pinging on this, I'm still seeing a mismatch between the flag name and flag doc string...?

Comment (ecumene) [reply crates/manifest/src/lib.rs:162]: I didn't see your reply:

I bounced the name `monitoring` back and forth with @lann. I suggested it but I don't think it's an accurate description either. I kept it hoping someone would come up with something better.

The flag name should be changed, I agree. How about something like `enable_home` or `mount_home`? I think someone with more experience than I could chime in

Comment (itowlson) [reply crates/manifest/src/lib.rs:162]: `mount_home_dir` or `access_home_dir` perhaps?  I'm going by the doc string, though, I don't really understand the feature, so I could be way off beam!  From the screenshot... this is allowing the hosted modules access to each others' asset directories?  In which case `access_other_component_dirs` might be clearer than "home"...?

Comment (itowlson) [reply crates/manifest/src/lib.rs:162]: (Many apologies if I am being obtuse here... I haven't followed enough and should have avoided suggesting things from a position of ignorance!)

Comment (lann) [reply crates/manifest/src/lib.rs:162]: The current iteration of this feature has a very straightforward effect: the `~/.spin/<your-app>/` dir is mounted in the component's WASI env as `/.spin/`. Currently the only useful files we expect there are the `logs/<component>_std{err,out}.txt`s, but anything that Spin wants to put in that dir would be available, which is why I'm fine with the more general "monitoring" term.

**Review by fibonacci1729 — APPROVED** [review]
